Send websocket events based on user attributes

jraddaoui opened the pull request:
- Allow saving key/value in the auth ticket store
- Use constants for API auth attributes/scopes
- Add auth.TicketProvider interface
- Send websocket events based on user attributes

Refs #1221.

jraddaoui committed:
Allow saving key/value in the auth ticket store

Gives the option to store a value associated with a generated ticket.
The ticket/key is sent as a cookie while the value is saved in the store
and scanned when the cookie is received. This will allow to save the
user claims between the monitor request and the monitor endpoints.

jraddaoui committed:
Add auth.TicketProvider interface

jraddaoui commented:
@djjuhasz, this is ready for an initial code review, just lacking tests for the monitor endpoints. It'll make more sense if you review commit by commit, thanks ;)!

djjuhasz commented on internal/api/auth/ticket_test.go:
```diff
@@ -44,14 +44,15 @@ func TestTicketProviderRequest(t *testing.T) {
 		provider := auth.NewTicketProvider(ctx, store, rander)
 
-		ticket, err := provider.Request(ctx)
+		ticket, err := provider.Request(ctx, value)
```
`ticket` here is shadowing the value of `ticket` from the parent scope.

djjuhasz commented on internal/api/auth/ticket_test.go:
```diff
@@ -44,14 +44,15 @@ func TestTicketProviderRequest(t *testing.T) {
-		ticket, err := provider.Request(ctx)
+		ticket, err := provider.Request(ctx, value)
 		assert.NilError(t, err)
 		assert.Equal(t, ticket, ticket)
```
This is always true due to the shadowing of `ticket`.

jraddaoui replied:
:man_facepalming:

djjuhasz commented on internal/api/auth/ticket_test.go:
```diff
@@ -67,7 +68,7 @@ func TestTicketProviderRequest(t *testing.T) {
-		ticket, err := provider.Request(ctx)
+		ticket, err := provider.Request(ctx, nil)
 		assert.Error(t, err, "error creating ticket: rand source error")
 		assert.Assert(t, ticket == "")
```
I'd expect `assert.Equal(t, ticket, "")` here and I think it will provide better reporting if the assertion fails.

djjuhasz commented on internal/api/auth/ticket_test.go:
```diff
@@ -81,13 +82,13 @@ func TestTicketProviderRequest(t *testing.T) {
-		ticket, err := provider.Request(ctx)
+		ticket, err := provider.Request(ctx, nil)
 		assert.Error(t, err, "error storing ticket: fake error")
 		assert.Assert(t, ticket == "")
```
Again, I think `assert.Equal()` is better here.

djjuhasz commented on internal/api/auth/ticket_test.go:
```diff
@@ -154,7 +161,7 @@ func TestTicketProviderCheck(t *testing.T) {
 		provider := auth.NewTicketProvider(ctx, nil, rander)
 
-		err := provider.Check(ctx, nil)
+		err := provider.Check(ctx, nil, nil)
```
What happens if you pass a ticket here?  Does the check always succeed if no store is defined?

jraddaoui replied:
Yes: https://github.com/artefactual-sdps/enduro/blob/main/internal/api/auth/ticket.go#L67

TBH, I never liked how we deal with API authentication being disabled, it's all kind of hidden in the token verifier and the ticket provider.

djjuhasz commented on internal/api/auth/attributes.go:
```diff
@@ -0,0 +1,25 @@
+package auth
+
+const (
+	IngestSipsDownloadAttr         = "ingest:sips:download"
```
According to our style guide "SIP" and "AIP" should be all caps: https://google.github.io/styleguide/go/decisions.html#initialisms

jraddaoui replied:
`IngestSIPsDownloadAttr` or `IngestSIPSDownloadAttr` :thinking:

jraddaoui replied:
I'll go with `IngestSIPSDownloadAttr`.

jraddaoui commented:
@djjuhasz, there is a "small" back and forward between https://github.com/artefactual-sdps/enduro/pull/1288/commits/e5df4344853e87f00f4f033e4427cf867dc417a0 and https://github.com/artefactual-sdps/enduro/pull/1288/commits/0e9b2aac6449187fc63a620fc9daac5dfa6e2476 that I realized after adding the tests.

This makes me wonder if we should just log the errors instead of breaking that loop (except context done).

jraddaoui commented:
Maybe that's the issue in https://github.com/artefactual-sdps/enduro/issues/1057, I'll take a look.

I'll squash the latest three commits ;)

djjuhasz commented on internal/ingest/monitor_test.go:
```diff
@@ -0,0 +1,247 @@
+			},
+		},
+		{
+			name: "Filters all events for a user wihtout permissions",
```
Typo: "wihtout"

djjuhasz reviewed (changes requested):
LGTM @jraddaoui, I just flagged some minor issues.

jraddaoui commented:
Thanks @djjuhasz, ready for another look!

djjuhasz reviewed (approved):
Looks great, thanks @jraddaoui! :+1:

jraddaoui committed:
Add noattributes user to Keycloak

jraddaoui committed:
Fix API auth ticket in memory store

Avoid nil pointer issue when the `[api.auth.ticket.redis]` section
is not included in the configuration.

jraddaoui committed:
Use constants for API auth attributes/scopes

jraddaoui committed:
Send websocket events based on user attributes

jraddaoui committed:
Fix websocket connection when auth is disabled

Load auth store configuration and immediatly watch if the user is valid
to connect to the ingest monitor when authentication is disabled.

jraddaoui commented:
Thanks @djjuhasz! I've squashed everything you already reviewed into the first four commits. But, while testing all possible permutations, I found a couple of issues that required small fixes. Please, take a look to the latest three commits when you have some time.

djjuhasz reviewed (approved):
The new changes look fine to me @jraddaoui.  It's not clear to me from the commits why you added a "noattributes" user, but it seems fine.

jraddaoui commented:
Thanks @djjuhasz, just a user that can authenticate but can't do anything, only for testing.
